### ISSUE TITLE
[Potential Bug] KL compute in `low_var_kl (Cause KL NaN and !!!!!!!!!!!!!!! output)

### ISSUE THREAD

**Issue by yushuiwx** [volcengine/verl#891]
Hi, 

A few days ago, I reported this issue in the discussion [751](https://github.com/volcengine/verl/issues/751) and now there are some updates.

similar issue:
- https://github.com/volcengine/verl/issues/747
- https://github.com/volcengine/verl/issues/721
-  .......

After careful analysis, I found that the issue lies in the calculation of `low_var_kl`. In the problematic step, the loss has normal values (without NaN or Inf), but during backpropagation, the `exp` operation in `ratio = torch.exp(kl)` within `low_var_kl` （https://github.com/volcengine/verl/blob/8cae42dc29736d0802ded43c5ecf67a809d56bd8/verl/trainer/ppo/core_algos.py#L386）causes extreme values, leading to NaN in `kl`. My temporary solution is to use `torch.clamp(kl, min=-5, max=5)` to constrain `kl` before computing `ratio`. **This approach works, but I feel uncertain about it. Could you kindly help confirm whether this is an appropriate solution?** 

 _Originally posted by @yushuiwx in [#751](https://github.com/volcengine/verl/issues/751#issuecomment-2774700868)_

**Comment by vermouth1992** [comment]
The final clamp does not take effect?

**Comment by great-luao** [comment]
For me I always have the similar !!! problem after the first training step. I found that my training grad_norm is 0, which caused the training to crash, I solve this issue refer to https://github.com/volcengine/verl/issues/405. Hope this will help somebody lol.
Besides, this leads to another issue that the assertion added actually won't work since we don't use "ppo_micro_batch_size" anymore. The assert code will be skipped and no one can notice.

![Image](https://github.com/user-attachments/assets/ff7850b6-1040-450e-bdbf-54b38fd77092)

**Comment by chuangzhidan** [comment]
> For me I always have the similar !!! problem after the first training step. I found that my training grad_norm is 0, which caused the training to crash, I solve this issue refer to [#405](https://github.com/volcengine/verl/issues/405). Hope this will help somebody lol. Besides, this leads to another issue that the assertion added actually won't work since we don't use "ppo_micro_batch_size" anymore. The assert code will be skipped and no one can notice.
> 
> ![Image](https://github.com/user-attachments/assets/ff7850b6-1040-450e-bdbf-54b38fd77092)

how did u fix it ?    
2gpus,my setting is like this,
actor_rollout_ref.actor.ppo_mini_batch_size=8 \
actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=2 \

**Comment by chuangzhidan** [comment]
> Hi,
> 
> A few days ago, I reported this issue in the discussion [751](https://github.com/volcengine/verl/issues/751) and now there are some updates.
> 
> similar issue:
> 
> * [训练会突然在某个步骤失败，模型某个指标会变成 nan，奖励变成0，然后开始不停重复生成感叹号 !!!!!!!!! #747](https://github.com/volcengine/verl/issues/747)
> * [output is  !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! #721](https://github.com/volcengine/verl/issues/721)
> * .......
> 
> After careful analysis, I found that the issue lies in the calculation of `low_var_kl`. In the problematic step, the loss has normal values (without NaN or Inf), but during backpropagation, the `exp` operation in `ratio = torch.exp(kl)` within `low_var_kl` （[https://github.com/volcengine/verl/blob/8cae42dc29736d0802ded43c5ecf67a809d56bd8/verl/trainer/ppo/core_algos.py#L386）causes](https://github.com/volcengine/verl/blob/8cae42dc29736d0802ded43c5ecf67a809d56bd8/verl/trainer/ppo/core_algos.py#L386%EF%BC%89causes) extreme values, leading to NaN in `kl`. My temporary solution is to use `torch.clamp(kl, min=-5, max=5)` to constrain `kl` before computing `ratio`. **This approach works, but I feel uncertain about it. Could you kindly help confirm whether this is an appropriate solution?**
> 
> _Originally posted by [@yushuiwx](https://github.com/yushuiwx) in [#751](https://github.com/volcengine/verl/issues/751#issuecomment-2774700868)


 i used torch.clamp(kl, min=-5, max=5),it didn't work ,sadlly

**Comment by Xia723** [comment]
依旧不行

**Comment by water-vapor** [comment]
Thanks for the solution, but it seems to address one consequence rather than the root cause. Even when it's constrained and no longer produces`nan`, the training still fails due to spikes in the `kl_loss`.

**Comment by chuangzhidan** [comment]
i turn off use_kl_loss ,and it's still not working,so i  switched to openrlhf

**Comment by lynnliu030** [comment]
> For me I always have the similar !!! problem after the first training step. I found that my training grad_norm is 0, which caused the training to crash, I solve this issue refer to [#405](https://github.com/volcengine/verl/issues/405). Hope this will help somebody lol. Besides, this leads to another issue that the assertion added actually won't work since we don't use "ppo_micro_batch_size" anymore. The assert code will be skipped and no one can notice.
> 
> ![Image](https://github.com/user-attachments/assets/ff7850b6-1040-450e-bdbf-54b38fd77092)

How did you solve it exactly again? Thanks!

**Comment by takagi97** [comment]
我用vllm 0.8.2 + 2节点跑的时候也遇到这个问题，经人提醒，将vllm的v1引擎关掉之后，一切正常了。
具体操作是，设置export VLLM_USE_V1=0，并且训练参数里面不要使用设置
        actor_rollout_ref.rollout.enforce_eager=False \
        actor_rollout_ref.rollout.free_cache_engine=False \
就可以了

**Comment by takagi97** [comment]
> > 我用vllm 0.8.2 + 2节点跑的时候也遇到这个问题，经人提醒，将vllm的v1引擎关掉之后，一切正常了。 具体操作是，设置export VLLM_USE_V1=0，并且训练参数里面不要使用设置 actor_rollout_ref.rollout.enforce_eager=False actor_rollout_ref.rollout.free_cache_engine=False 就可以了
> 
> 请问这个是修复了什么问题？我现在是一直到平台，grad norm 会变成nan

我没有仔细研究过，我看到过别的issue讨论过这种突然训练爆炸有一种成因是使用当前policy和old policy计算采样出来的sample的概率的差别过大，导致算出来的ratio数值爆炸。不清楚是不是这个原因，但是我这边v0能跑v1跑不了，估计是v1的bug可能引起上述问题。

**Comment by HollrayChan** [comment]
> 我用vllm 0.8.2 + 2节点跑的时候也遇到这个问题，经人提醒，将vllm的v1引擎关掉之后，一切正常了。 具体操作是，设置export VLLM_USE_V1=0，并且训练参数里面不要使用设置 actor_rollout_ref.rollout.enforce_eager=False actor_rollout_ref.rollout.free_cache_engine=False 就可以了

我这边4卡24g的，只设置actor_rollout_ref.rollout.enforce_eager=False 和actor_rollout_ref.rollout.free_cache_engine=False 就有效

**Comment by HardenGale** [comment]
same problem

**Comment by Kirrito-k423** [comment]
中文：我建议这个 issue 先保持打开并补充复现信息，而不是现在关闭。当前 main 的 `low_var_kl`/`k3` 路径已经在 `torch.exp(kl)` 前做了 `torch.clamp(kl, min=-20, max=20)`，并且对输出的 `kld` 做了 `[-10, 10]` clamp。因此，原始建议的“在 exp 前 clamp”方向在当前代码里已经存在，但评论中仍有用户反馈 clamp 后训练失败，另有用户反馈关闭 vLLM V1 或调整 `enforce_eager`/`free_cache_engine` 后可缓解。

为了继续定位，请受影响的用户补充：1. 使用的 verl commit 或 release 版本；2. vLLM 版本以及是否设置 `VLLM_USE_V1=0`；3. 完整训练命令和关键配置；4. NaN 前后 2-3 个 step 的 `old_log_prob`、`log_prob`、`kl_loss`、`ppo_kl`、`grad_norm` 日志；5. 是否在最新 main 上尝试过 TIS 相关配置或 `low_var_kl+`。有了这些信息后，才能判断这是 KL estimator 的数值稳定性问题、rollout/training log-prob mismatch，还是特定 vLLM/配置组合导致的问题。

English: I suggest keeping this issue open and asking for more reproduction details instead of closing it now. On current main, the `low_var_kl`/`k3` path already clamps `kl` to `[-20, 20]` before `torch.exp(kl)`, and also clamps the final `kld` to `[-10, 10]`. So the original “clamp before exp” direction is already present in current code, but comments show mixed results: some users still fail after clamping, while others report that disabling vLLM V1 or changing `enforce_eager`/`free_cache_engine` helps.

For further debugging, affected users should provide: 1. the verl commit or release version; 2. the vLLM version and whether `VLLM_USE_V1=0` is set; 3. the full training command and key configs; 4. `old_log_prob`, `log_prob`, `kl_loss`, `ppo_kl`, and `grad_norm` logs for 2-3 steps before and after NaN; 5. whether they have tried the latest main with TIS-related configs or `low_var_kl+`. With that information, maintainers can tell whether this is a KL estimator numerical stability issue, a rollout/training log-prob mismatch, or a vLLM/config-specific failure.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
